webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/reports/ExecutionBatch.scala

```diff
@@ -831,6 +841,168 @@ final case class ContextForNoAnswer(
+    val t1 = System.nanoTime
+    //here, we had at least one report, even if it not a ResultReports (i.e: run start/end is meaningful
+
+    val reports = reportsForThatNodeRule.groupBy(x => (x.directiveId, x.component) )
+
+    val expectedComponents: Map[(DirectiveId, List[EffectiveExpectedComponent]), (PolicyMode, ReportType)] = (for {
+      directive  <- directives
+      policyMode =  PolicyMode.directivePolicyMode(
+                           modes.globalPolicyMode
+                         , modes.nodePolicyMode
+                         , directive.policyMode
+                         , directive.isSystem
+                    )
+      // the status to use for ACTUALLY missing reports, i.e for reports for which
+      // we have a run but are not here. Basically, it's "missing" when on
+      // full compliance and "success" when on changes only - but that success
+      // depends upon the policy mode
+      missingReportStatus = missingReportType(modes.globalComplianceMode, policyMode)
+
+      component  <- directive.components
+
+    } yield {
+
+      ((directive.directiveId, getExpectedComponents(component)), (policyMode, missingReportStatus))
+    }).toMap
+    val t2 = System.nanoTime
+    timer.u1 += t2-t1
+
+    /*
+     * now we have three cases:
+     * - expected component without reports => missing (modulo changes only interpretation)
+     * - reports without expected component => unknown
+     * - both expected component and reports => check
+     */
+    val reportKeys = reports.keySet
+    val expectedKeys: Set[(DirectiveId,String)] = expectedComponents.keySet.flatMap(c => c._2.map(d => (c._1, d.value.componentName)))
+    val okKeys = reportKeys.intersect(expectedKeys)
+
+    val t3 = System.nanoTime
+    timer.u2 += t3-t2
+
+    //unexpected contains the one with unexpected key and all non matching serial/version
+    val unexpected = (if (okKeys.size != reportKeys.size) {
+      buildUnexpectedDirectives(
+        reports.filter(k => !expectedKeys.contains((k._1._1,k._1._2)                                       )).values.flatten.toSeq
+      )
+    } else {
+      Seq[DirectiveStatusReport]()
+    })
+
+    if (ruleId == RuleId("32377fd7-02fd-43d0-aab7-28460a91347b")) logger.error(unexpected)
+    val t4 = System.nanoTime
+    timer.u3 += t4-t3
+
+    // okKeys is DirectiveId, ComponentName
+    val expected: Iterable[DirectiveStatusReport] =
+      expectedComponents.groupBy(_._1._1).map {
+        case (directiveId, expectedComponentsForDirective) =>
+          DirectiveStatusReport(directiveId, expectedComponentsForDirective.flatMap {
+            case ((directiveId, components), (policyMode, missingReportStatus)) =>
+              val r = components.map { c =>
+                val filteredReports = reports.flatMap { case ((id, cname), r) => r.filter(value => directiveId == id && cname == c.value.componentName && c.value.componentsValues.exists(v => checkExpectedVariable(v, value.keyValue))) }.toList
```

Comment:
unless i'm mistaken, we have a quadratic complexity on c.value.componentsValues.exists(v => checkExpectedVariable(v, value.keyValue))
since it's fairly expensive to do the `matchCFEngineVars.pattern.matcher(expected).matches()`, it could be precomputed or cached once for all values